tests/test_paper.py

```diff
@@ -0,0 +1,209 @@
+                # This is actually a syncronous call to push a message
+                # to the async ems clue - hence why we call trio.sleep afterwards
+                book.send(order)
+
+                async for msg in trades_stream:
+                    last_msg = msg
+                    match msg:
+                        # Wait for position message before moving on
+                        case {'name': 'position'}:
+                            break
+
+        if assert_entries or assert_pps or assert_zeroed_pps or assert_msg:
+            _assert(
+                assert_entries,
+                assert_pps,
+                assert_zeroed_pps,
+                pps,
+                last_msg,
+                size,
+                executions,
+            )
+
+        # Teardown piker like a user would
+        raise KeyboardInterrupt
+
+
+def _assert(
+    assert_entries,
+    assert_pps,
+    assert_zerod_pps,
+    pps,
+    last_msg,
+    size,
+    executions,
+):
+    with (
+        open_trade_ledger(broker, account) as ledger,
+        open_pps(broker, account) as table,
+    ):
+        '''
+        Assert multiple cases including pps, ledger and final position message state
+        '''
+        if assert_entries:
+            assert last_msg['broker'] == broker
+            assert last_msg['account'] == account
+            assert last_msg['symbol'] == fqsn
+            assert last_msg['size'] == size * executions
+            assert last_msg['currency'] == symbol
+            assert last_msg['avg_price'] == table.pps[symbol].ppu
+
+        if assert_pps:
+            last_ppu = pps[(broker, account)][-1]
+            assert last_ppu['avg_price'] == table.pps[symbol].ppu
+
+        if assert_zerod_pps:
+            assert not bool(table.pps)
+
+
+# Close position and assert empty position in pps
+def _run_test_and_check(fn):
+    with pytest.raises(BaseExceptionGroup) as exc_info:
+        trio.run(fn)
+
+    for exception in exc_info.value.exceptions:
+        assert isinstance(exception, KeyboardInterrupt) or isinstance(
+            exception, ContextCancelled
+        )
+
+
+def test_buy(open_test_pikerd_and_ems: AsyncContextManager, delete_testing_dir):
+    # Enter a trade and assert entries are made in pps and ledger files
+    _run_test_and_check(
+        partial(
+            _async_main,
+            open_test_pikerd_and_ems=open_test_pikerd_and_ems,
+            action='buy',
+            assert_entries=True,
+        ),
+    )
+
+    # Open ems and assert existence of pps entries
+    _run_test_and_check(
+        partial(
+            _async_main,
+            open_test_pikerd_and_ems=open_test_pikerd_and_ems,
+            assert_pps=True,
+        ),
+    )
+
+
+def test_sell(open_test_pikerd_and_ems: AsyncContextManager, delete_testing_dir):
+    # Sell position
+    _run_test_and_check(
+        partial(
+            _async_main,
+            open_test_pikerd_and_ems=open_test_pikerd_and_ems,
```

Comment:
2lines between top level defined functions as per my comment referring to `pep8` a while back.